### DISCUSSION TITLE
Next.js inside of Docker (with nginx) causes page to refresh every few minutes

### DISCUSSION THREAD

**Discussion by WKampel** [vercel/next.js#74287] [Help] [answered]
### Link to the code that reproduces this issue

https://github.com/WKampel/next-docker-refresh-issue-example

### To Reproduce

1. Clone the repository https://github.com/WKampel/next-docker-refresh-issue-example
2. Run docker-compose up
3. Open http://localhost:5050 in your web browser
4. The page will keep refreshing every 2-3 minutes. You can type text in the input field and notice it disappear on reload.

### Current vs. Expected behavior

The page shouldn't refresh every few minutes.

### Provide environment information

```bash
OS: Mac arm64
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Which stage(s) are affected? (Select all that apply)

next dev (local)

### Additional context

This issue does not occur on my production builds. Just development. Next.js does work outside of Docker. I'm not sure if this is a Next.js or Docker issue, but this never used to occur until last month (randomly). I haven't had similar issues with Docker and other frontend frameworks.

I reproduced this issue by following the reproduction instructions on my second mac.

**Comment by WKampel** [comment, marked as answer]
For those with the same issue, I finally solved it.

Turns out to be an issue with NGINX. I noticed in the console I was getting an error:

`WebSocket connection to 'ws://localhost:3050/_next/webpack-hmr' failed:`

I added code to my nginx config to fix that error, which inadvertently fixed the refresh problem.

 ```
 location /_next/webpack-hmr {
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header Host $host;
    proxy_pass http://frontend;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
  }
```

**Comment by Mluocheng** [comment]
应该是部署的问题， 我测试在服务器的web应用使用npm run dev 回遇到这个问题， 将项目打包后 使用pm2 来启动，就没有这个问题。 且可以看到 有两个应用被启动了。
![image](https://github.com/vercel/next.js/assets/31057183/c9b88bee-97d6-4021-85ba-bbf37f1db1eb)
猜测next 有socket 服务需要启动。

**Comment by EliasBobadilla** [comment]
> For those with the same issue, I finally solved it.
> 
> Turns out to be an issue with NGINX. I noticed in the console I was getting an error:
> 
> `WebSocket connection to 'ws://localhost:3050/_next/webpack-hmr' failed:`
> 
> I added code to my nginx config to fix that error, which inadvertently fixed the refresh problem.
> 
> ```
> location /_next/webpack-hmr {
>    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
>    proxy_set_header Host $host;
>    proxy_pass http://frontend;
>    proxy_http_version 1.1;
>    proxy_set_header Upgrade $http_upgrade;
>    proxy_set_header Connection "upgrade";
>  }
> ```

Thank you so much for sharing. It worked perfectly for me. Just a nit: the `proxy_pass` needs the port explicit in my case `proxy_pass http://web:${WEB CLIENT_PORT};`

**Comment by weee-shinan** [comment]
> 应该是部署的问题， 我测试在服务器的web应用使用npm run dev 回遇到这个问题， 将项目打包后 使用pm2 来启动，就没有这个问题。 且可以看到 有两个应用被启动了。 ![image](https://private-user-images.githubusercontent.com/31057183/313384689-c9b88bee-97d6-4021-85ba-bbf37f1db1eb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUwMzUyNDIsIm5iZiI6MTczNTAzNDk0MiwicGF0aCI6Ii8zMTA1NzE4My8zMTMzODQ2ODktYzliODhiZWUtOTdkNi00MDIxLTg1YmEtYmJmMzdmMWRiMWViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI0VDEwMDkwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFmOTgwMDk0MGU5YzlmMThkNzgzYWNlYmMwMDM2YjFlZmM0ODAwOWZkMzExMjRiZjM5NGU1YTJmNGU1NzQzMDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.BxacrOEf_hB8aVC3jrZSrBGP4PvreYqtNgQCwJZxkEI) 猜测next 有socket 服务需要启动。

我遇到的问题跟你相反, 我想使用 pm2 来启动 nextjs 项目, Docker 跑起来之后, 访问页面每几秒就刷新一次, 发现是有 _ssgManifest.js 文件500, 很奇怪
